auth/src/krb.cpp

```diff
@@ -0,0 +1,942 @@
+#include <irods/irods_auth_constants.hpp>
+#include <irods/irods_auth_plugin.hpp>
+#include <irods/irods_client_server_negotiation.hpp>
+#include <irods/irods_configuration_keywords.hpp>
+#include <irods/irods_error.hpp>
+#include <irods/irods_krb_object.hpp>
+#include <irods/irods_kvp_string_parser.hpp>
+#include <irods/irods_rs_comm_query.hpp>
+#include <irods/irods_server_properties.hpp>
+#include <irods/irods_stacktrace.hpp>
+#include <irods/miscServerFunct.hpp>
+#include <irods/rodsErrorTable.h>
+#include <irods/rodsLog.h>
+
+#ifdef RODS_SERVER
+#include <irods/rsGenQuery.hpp>
+#include <irods/rsAuthCheck.hpp>
+#include <irods/rsAuthResponse.hpp>
+#include <irods/rsAuthRequest.hpp>
+#endif // RODS_SERVER
+
+#include <gssapi.h>
+#include <openssl/md5.h>
+
+#include <string>
+
+static const int ikrbDebugFlag = 0;
+static const gss_OID gss_nt_service_name_krb = 0;
+static const unsigned int SCRATCH_BUFFER_SIZE = 20000;
+static char ikrbScratchBuffer[SCRATCH_BUFFER_SIZE];
+static int ikrbTokenHeaderMode = 1;  /* 1 is the normal mode,
+                                        0 means running in a non-token-header mode, ie Java; dynamically cleared. */
+
+#ifdef RODS_SERVER
+// =-=-=-=-=-=-=-
+// NOTE:: this needs to become a property
+// Set requireServerAuth to 1 to fail authentications from
+// un-authenticated Servers (for example, if the LocalZoneSID
+// is not set)
+//static const int requireServerAuth = 0;
+static int krbAuthReqStatus = 0;
+static int krbAuthReqError = 0;
+static const int krbAuthErrorSize = 1000;
+static char krbAuthReqErrorMsg[krbAuthErrorSize];
+#endif
+
+
+static rError_t *ikrb_rErrorPtr;
+
+// static gss_cred_id_t myCreds = GSS_C_NO_CREDENTIAL;
+static const int MAX_FDS = 32;
+static gss_ctx_id_t context[MAX_FDS] = {
+    GSS_C_NO_CONTEXT, GSS_C_NO_CONTEXT, GSS_C_NO_CONTEXT, GSS_C_NO_CONTEXT,
+    GSS_C_NO_CONTEXT, GSS_C_NO_CONTEXT, GSS_C_NO_CONTEXT, GSS_C_NO_CONTEXT,
+    GSS_C_NO_CONTEXT, GSS_C_NO_CONTEXT, GSS_C_NO_CONTEXT, GSS_C_NO_CONTEXT,
+    GSS_C_NO_CONTEXT, GSS_C_NO_CONTEXT, GSS_C_NO_CONTEXT, GSS_C_NO_CONTEXT,
+    GSS_C_NO_CONTEXT, GSS_C_NO_CONTEXT, GSS_C_NO_CONTEXT, GSS_C_NO_CONTEXT,
+    GSS_C_NO_CONTEXT, GSS_C_NO_CONTEXT, GSS_C_NO_CONTEXT, GSS_C_NO_CONTEXT,
+    GSS_C_NO_CONTEXT, GSS_C_NO_CONTEXT, GSS_C_NO_CONTEXT, GSS_C_NO_CONTEXT,
+    GSS_C_NO_CONTEXT, GSS_C_NO_CONTEXT, GSS_C_NO_CONTEXT, GSS_C_NO_CONTEXT
+};
+
+static unsigned int context_flags;
+
+static gss_cred_id_t my_creds = GSS_C_NO_CREDENTIAL;
+
+namespace
+{
+    using json = nlohmann::json;
+    using log_auth = irods::experimental::log::authentication;
+
+    auto request(rcComm_t& comm, const json& msg)
+    {
+        auto str = msg.dump();
+
+        bytesBuf_t inp;
+        inp.buf = static_cast<void*>(const_cast<char*>(str.c_str()));
+        inp.len = str.size();
+
+        bytesBuf_t* resp{};
+        auto ec = procApiRequest(&comm, 110000, static_cast<void*>(&inp), nullptr, reinterpret_cast<void**>(&resp), nullptr);
```

Comment:
Consider giving `110000` a name?

Reply:
Ah, I thought I'd done this with the other plugins/the original port of the AWG effort, but I guess not. This convenience function is being used in all of the new auth plugins, so it should be pulled into a common library anyway. I think we can modify the API number at that point, if that's okay.

Reply:
I have added this to the list of tasks for wrapping up #5479: https://github.com/irods/irods/issues/5479#issuecomment-1115234800

Reply:
Done in https://github.com/irods/irods/pull/6358 and removed here